src/System.Memory/tests/MemoryMarshal/CreateFromPinnedArray.cs

```diff
@@ -0,0 +1,160 @@
+            memory.Validate(42, -1);
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithNegativeStartAndLength()
+        {
+            int[] a = new int[3];
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, -1, 0));
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithStartTooLargeAndLength()
+        {
+            int[] a = new int[3];
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 4, 0));
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithStartAndNegativeLength()
+        {
+            int[] a = new int[3];
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 0, -1));
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithStartAndLengthTooLarge()
+        {
+            int[] a = new int[3];
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 3, 1));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 2, 2));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 1, 3));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 0, 4));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, int.MaxValue, int.MaxValue));
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithStartAndLengthBothEqual()
+        {
+            // Valid for start to equal the array length. This returns an empty memory that starts "just past the array."
+            int[] a = { 91, 92, 93 };
+            Memory<int> pinnedMemory = MemoryMarshal.CreateFromPinnedArray(a, 3, 0);
+            pinnedMemory.Validate();
+        }
+
+        [Fact]
+        public static unsafe void CreateFromPinnedArrayVerifyPinning()
+        {
+            int[] unpinnedArray = { 90, 91, 92, 93, 94, 95, 96, 97, 98 };
+            int[] pinnedArray = { 90, 91, 92, 93, 94, 95, 96, 97, 98 };
+            GCHandle handle = GCHandle.Alloc(pinnedArray, GCHandleType.Pinned);
+
+            var memory = new Memory<int>(unpinnedArray, 0, 2);
+            void* ptr = Unsafe.AsPointer(ref MemoryMarshal.GetReference(memory.Span));
+
+            MemoryHandle memoryHandle = memory.Pin();
+            void* memoryHandlePtr = memoryHandle.Pointer;
+            memoryHandle.Dispose();
+
+            Memory<int> pinnedMemory = MemoryMarshal.CreateFromPinnedArray(pinnedArray, 0, 2);
+            void* pinnedPtr = Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedMemory.Span));
+            void* memoryHandlePinnedPtr = pinnedMemory.Pin().Pointer;
+
+            GC.Collect();
+            GC.Collect(2);
```

Comment:
I don't think this is absolutely guaranteeing that the GC will move the array, and so the test might fail from time to time.

Reply:
That's what I thought. In which case, I will remove this test (but I can keep the "pinned" components).